tools/Dockerfile

```diff
@@ -0,0 +1,34 @@
+#
+# Dockerfile for quickstart containerised scipy development
+#
+#
+# Usage:
+# ------
+# If you make changes to the Dockerfile you can rebuild the container:
+#
+# docker build .
+# docker push <build-hash> scipy/scipy-dev
+#
+# To use the container use the following command. It assumes that you are in a
+# directory containing the scipy git repository, mounting the scipy directory
+# in $PWD and making it available as /home/scipy in the container. Whatever
+# changes you make to that directory are visible in the host and container.
+# The docker image is retrieved from the scipy dockerhub repository
+#
+# docker run -it --rm -v $PWD/scipy:/home/scipy scipy/scipy-dev /bin/bash
+#
+# The available Python executables are python3.6 and python3.7 (along with
+# pip3.6 and pip3.7.
+# The image does not come bundled with cython/numpy/pytest, those need to
+# be installed with pip before trying to build/run scipy.
```

Comment:
why not add this? one of the advantages of using Docker is that things get fullly reproducible, and that's sometimes worth the extra hassle. so may as well make an image that's 100% complete

Reply:
It depends on how the Dockerfile will be used. 

Case 1 - Dockerfile is used directly by the end user
----------------------------------------------------
Here the user rebuilds the image everytime that want to start the container. Docker caches various things using a layer approach - this [article is worth reading](https://medium.com/@jessgreb01/digging-into-docker-layers-c22f948ed612). Adding a pip install of numpy/cython/pytest/pybind11 makes sense here as the container image (virtual machine) is being built each time, and amending the script to include necessary dependencies takes one step away from the user having to do it when they start the container. Building the image can take some time, and also bandwidth as all the dependencies have to be pulled in.

Case 2 - Dockerfile is used to make container image uploaded to Dockerhub
-----------------------------------------------------------------------------
scipy maintainers update this Dockerfile as necessary, build the image, and upload to Dockerhub (Dockerhub also has github hooks that allow the image to be built automatically). The end user runs the image with a single command, `docker run scipy/scipy-dev`. If the image is stored locally then the docker container starts immediately, otherwise docker will fetch the image from the scipy image repo on dockerhub. It's much faster to get an image than build an image. The cheapness of starting an image is why it's so good for CI. If the user is trying to do something like test a matrix of numpy versions then they would have to uninstall the version that came with the image, then reinstall the version they wanted. 
  - That uninstall/reinstall step is one of the main reasons. 
  - I based the Dockerfile on what we currently do on TravisCI; all of the Linux matrix entries are common up to the point where the various numpys and other test options start the differentiation. I was aiming for that point. 
  - Lastly, the built image as it currently exists has a reasonable good shelf life (modulo Python updates) for a while. If we put cython/numpy/pytest/pybind11 on the image then the shelf life of the image is only as long as it takes for another numpy update to come out.

Reply:
The shelf life is a good point. It could be addressed with a separate image for the less stable components though. I think the problem is now that, with case 2, a user has to reinstall `numpy`, `cython` et al. every time the image is run.

Reply:
Also, more dependencies are missing: matplotlib, Sphinx, scikit-umfpack (perhaps more, didn't check the CI scripts).

Reply:
> how about running a script automatically at startup to install the most recent version of dependencies?

Not really a fan of that, I think it's worse than the main two options.